Simulator/actions/messaging_actions.js

```diff
@@ -0,0 +1,257 @@
+          break;
+        }
+
+        case eventNames.ABORT_AUTH: {
+          const errorMsg = msgData.errorMsg;
+          dispatch(handleAbortForAuth(errorMsg));
+          break;
+        }
+        default: {
+          // The message was not for the simulator, ignore it.
+          break;
+        }
+      }
+    } catch (e) {
+      // The message was not for the simulator, ignore it.
+      return;
+    }
+  };
+}
+
+export function handleLog(msg) {
+  return () => {
+    /* eslint-disable no-console */
+    console.log(msg);
+    /* eslint-enable no-console */
+  };
+}
+
+export function handleLoaded() {
+  return (dispatch) => {
+    dispatch(sendInit());
+  };
+}
+
+export function handleInit() {
+  return (dispatch, getState) => {
+    const { currentPhase } = getState();
+    dispatch(simulatorActions.setPhaseInitCallbackCalled(true));
+    dispatch(simulatorActions.setPhaseInProgress(true));
+    // If we are in the Gather Data Phase, ask wdc for headers
+    if (currentPhase === phases.GATHER_DATA) {
+      dispatch(sendGetHeaders());
+    }
+  };
+}
+
+export function handleSubmit() {
+  return (dispatch) => {
+    // Clean up simulator and start Data Gather Phase
+    dispatch(simulatorActions.setPhaseSubmitCalled(true));
+    dispatch(simulatorActions.setPhaseInProgress(false));
+    dispatch(simulatorActions.closeSimulatorWindow());
+    dispatch(simulatorActions.resetTables());
+    dispatch(simulatorActions.startGatherDataPhase());
+  };
+}
+
+export function handleSchemaCallback(schema) {
+  return (dispatch, getState) => {
+    // Validate schema, and populate store with new table objects
+    // using the schema info
+    const { wdcShouldFetchAllTables } = getState();
+    if (validateSchema(schema)) {
+      const newTables = schema.reduce((tables, tableInfo) => {
+        const newTable = {
+          schema: tableInfo,
+          data: [],
+        };
+
+        return { ...tables, [tableInfo.id]: newTable };
+      }, {});
+      dispatch(simulatorActions.addTables(newTables));
+
+      // if we are supposed to automatically fetch the tables, do it
+      if (wdcShouldFetchAllTables) {
+        dispatch(fetchAllData());
+      } else {
+        // We we aren't going to fetch immediately
+        // consider the phase paused until the getGeta
+        // message is sent. This is so we can properly
+        // disable the Fetch Data button
+        dispatch(simulatorActions.setPhaseInProgress(false));
+      }
+    } else {
+      toastr.error('Please see debug console for details.', 'WDC Validation Error');
+    }
+  };
+}
+
+export function handleDataCallback(tableName, data) {
+  return (dispatch, getState) => {
+    // if data is valid, add it to the appropriate tables object in the store
+    if (validateData(data)) {
+      const { tables } = getState();
+      let newTables = Object.assign({}, tables);
+      newTables[tableName].data = newTables[tableName].data.concat(data);
```

Comment:
We can have data passed back either as an array or arrays, or as an array of objects. Does this cause any issues here? (I haven't read where we read this data yet)